File: mobile/lib/protocol/ack_payload.dart
import 'dart:typed_data';

import 'payload_codec.dart';

class AckPayload {
  const AckPayload({required this.ackedSequence, required this.ackTime});

  final int ackedSequence;
  final int ackTime;

  @override
  bool operator ==(Object other) =>
      other is AckPayload &&
      ackedSequence == other.ackedSequence &&
      ackTime == other.ackTime;

  @override
  int get hashCode => Object.hash(ackedSequence, ackTime);
}

class AckPayloadCodec {
  static Uint8List encode(AckPayload payload) {
    final writer = PayloadWriter()
      ..writeUInt16(payload.ackedSequence)
      ..writeUInt64(payload.ackTime);
    return writer.toBytes();
  }

  static AckPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final ackedSequence = reader.readUInt16();
    final ackTime = reader.readUInt64();
    reader.expectEnd();
    return AckPayload(ackedSequence: ackedSequence, ackTime: ackTime);
  }
}

File: desktop/Protocol/AckPayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record AckPayload(ushort AckedSequence, ulong AckTime);

public static class AckPayloadCodec
{
    public static byte[] Encode(AckPayload payload)
    {
        var writer = new PayloadWriter();
        writer.WriteUInt16(payload.AckedSequence);
        writer.WriteUInt64(payload.AckTime);
        return writer.ToArray();
    }

    public static AckPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var result = new AckPayload(reader.ReadUInt16(), reader.ReadUInt64());
        reader.ExpectEnd();
        return result;
    }
}
